### ISSUE TITLE
关于训练效果的一些疑问

### ISSUE THREAD

**Issue by apoca909** [yongzhuo/macro-correct#19]
训练模型的相关参数如下：
{
    "pretrained_model_name_or_path": "hfl/chinese-macbert-base",
    "path_train": "D:/workspace/MateRepos/mate_csc/datasets/csc/sighan2015/train3.txt",
    "path_dev": "D:/workspace/MateRepos/mate_csc/datasets/csc/sighan2015/test_merge.txt",
    "path_tet": "D:/workspace/MateRepos/mate_csc/datasets/csc/sighan2015/test_merge.txt",
    "model_save_path": "../../../output/text_correction\\macbert4mdcspell",
    "task_name": "macbert4mdcspell",
    "flag_fast_tokenizer": true,
    "do_lower_case": true,
    "do_train": true,
    "do_eval": true,
    "do_test": true,
    "gradient_accumulation_steps": 4,
    "warmup_proportion": 0.1,
    "num_warmup_steps": null,
    "max_train_steps": 123390,
    "num_train_epochs": 10,
    "train_batch_size": 8,
    "eval_batch_size": 16,
    "learning_rate": 3e-05,
    "max_seq_length": 128,
    "max_grad_norm": 1.0,
    "weight_decay": 0.0005,
    "save_steps": 6169,
    "anchor": null,
    "seed": 42,
    "lr_scheduler_type": "linear",
    "loss_type": "focal_loss",
    "mask_mode": "noerror",
    "loss_det_rate": 0.15,
    "prompt_length": 0,
    "mask_rate": 0.2,
    "threshold": 0.5,
    "flag_dynamic_encode": false,
    "flag_loss_period": false,
    "flag_cpo_loss": false,
    "flag_pin_memory": true,
    "flag_train": false,
    "flag_fp16": false,
    "flag_cuda": true,
    "flag_skip": true,
    "flag_mft": true,
    "num_workers": 0,
    "CUDA_VISIBLE_DEVICES": "0",
    "USE_TORCH": "1",
    "platform": "windows"
}
训练数据为wang271k+sighan+Medical、Lemon、ECSpell、CSCD 共计约39w条，训练10个epoch

部分效果如下：

11/05/2025 14:29:40 - INFO - __main__ -     Num examples = 49812
11/05/2025 14:29:40 - INFO - __main__ -     Batch size = 16
11/05/2025 14:36:12 - INFO - __main__ -   flag_eval: common
11/05/2025 14:36:12 - INFO - __main__ -   Sentence Level detection: acc:0.6829, precision:0.9131, recall:0.4966, f1:0.6433
11/05/2025 14:36:12 - INFO - __main__ -   Sentence Level correction: acc:0.6320, precision:0.8962, recall:0.4082, f1:0.5609
11/05/2025 14:36:13 - INFO - __main__ -   flag_eval: strict
11/05/2025 14:36:13 - INFO - __main__ -   Sentence Level detection: acc:0.6829, precision:0.5836, recall:0.4966, f1:0.5366
11/05/2025 14:36:13 - INFO - __main__ -   Sentence Level correction: acc:0.6320, precision:0.4797, recall:0.4082, f1:0.4410
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  cor_acc = 0.6319963061109772
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  cor_f1 = 0.4410329823503928
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  cor_precision = 0.47965747531445896
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  cor_recall = 0.4081653999023778
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  det_acc = 0.6829278085601863
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  det_f1 = 0.5366083369436231
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  det_precision = 0.5836030646945548
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  det_recall = 0.4966180879994422
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  eval_cor_loss = 0.046067647509563985
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  eval_cpo_loss = 0.09260262119588271
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  eval_det_loss = 0.0939158178770236
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  eval_loss = 0.056665299352360095
11/05/2025 14:36:13 - INFO - __main__ -   Global step: -1,  lr = 0.0

predict.py预测结果很难看：

{'source': '真麻烦你了。希望你们好好的跳无', 'target': '##橢##紅##笑炊##橢##疯##穷葩##笑対##呱ads##free##污##橢', 'errors': [['真', '#', 0], ['麻', '#', 1], ['烦', '橢', 2], ['你', '#', 3], ['了', '#', 4], ['。', '紅', 5], ['希', '#', 6], ['望', '#', 7], ['你', '笑', 8], ['们', '炊', 9], ['好', '#', 10], ['好', '#', 11], ['的', '橢', 12], ['跳', '#', 13], ['无', '#', 14]]}
{'source': '少先队员因该为老人让坐', 'target': '##橢##笑##卖##疯contact##free##橢胶##free##菖卯', 'errors': [['少', '#', 0], ['先', '#', 1], ['队', '橢', 2], ['员', '#', 3], ['因', '#', 4], ['该', '笑', 5], ['为', '#', 6], ['老', '#', 7], ['人', '卖', 8], ['让', '#', 9], ['坐', '#', 10]]}
{'source': '机七学习是人工智能领遇最能体现智能的一个分知', 'target': '##橢##︿##卖鱗憂##哌##穷##笑##橢##free##橢rose##體ok##哌##頷##橢##free##®station##橢mmorpg', 'errors': [['机', '#', 0], ['七', '#', 1], ['学', '橢', 2], ['习', '#', 3], ['是', '#', 4], ['人', '︿', 5], ['工', '#', 6], ['智', '#', 7], ['能', '卖', 8], [' 领', '鱗', 9], ['遇', '憂', 10], ['最', '#', 11], ['能', '#', 12], ['体', '哌', 13], ['现', '#', 14], ['智', '#', 15], ['能', '穷', 16], ['的', '#', 17], ['一', '#', 18], ['个', '笑', 19], ['分', '#', 20], ['知', '#', 21]]}
{'source': '一只小鱼船浮在平净的河面上', 'target': '卯amgwiidcs##習柵描鳄##橢670##垮wii##橢', 'errors': [['一', '卯', 0], ['只', 'a', 1], ['小', 'm', 2], ['鱼', 'g', 3], ['船', 'w', 4], ['浮', 'i', 5], ['在', 'i', 6], ['平', 'd', 7], ['净', 'c', 8], ['的', 's', 9], ['河', '#', 10], ['面', '#', 11], ['上', '習', 12]]}      
{'source': '我的家乡是有明的渔米之乡', 'target': '##橢##free##氟##穷憂##橢绦胶##挚対##橢##橢', 'errors': [['我', '#', 0], ['的', '#', 1], ['家', '橢', 2], ['乡', '#', 3], ['是', '#', 4], ['有', 'f', 5], ['明', 'r', 6], ['的', 'e', 7], ['渔', 'e', 8], ['米', '#', 9], ['之', '#', 10], ['乡', '氟', 11]]}
{'source': 'chuxī到了,兰兰和妈妈一起包jiǎo子,兰兰包的jiǎo子十分可爱,\n今天,兰兰和妈妈dù过了一个&快乐的chúxī。', 'target': '卯##嬗废雜##には##橢##４##麺義h2##娟wii鋤mama鋤##體##free##には##體＊##销##ハdcsdcs##free##呱##free杆##橢##笑##畏194anna##橢##彈##彈h2##畏amg##笑穎vetements篩亩##橢##卖柵##橢##橢##穷##橢##垮##橢h2amg描##摆##祉##橢穎', 'errors': [['c', '卯', 0], ['h', '#', 1], ['u', '#', 2], ['x', '嬗', 3], ['ī', '废', 4], ['到', '雜', 5], ['了', '#', 6], [',', '#', 7], ['兰', 'に', 8], ['兰', 'は', 9], ['和', '#', 10], ['妈', '#', 11], ['妈', '橢', 12], ['一', '#', 13], ['起', '#', 14], ['包', '４', 15], ['j', '#', 16], ['i', '#', 17], ['ǎ', '麺', 18], ['o', '義', 19], ['子', 'h', 20], [',', '2', 21], ['兰', '#', 22], ['兰', '#', 23], ['包', '娟', 24], ['的', 'w', 25], ['j', 'i', 26], ['ǎ', '鋤'['ǎ', '麺', 18], ['o', '義', 19], ['子', 'h', 20], [',', '2', 21], ['兰', '#', 22], ['兰', '#', 23], ['包', '娟', 24], ['的', 'w', 25], ['j', 'i', 26], ['ǎ', '鋤', 28], ['o', 'm', 29], ['子', 'a', 30], ['十', 'm', 31], ['分', 'a', 32], ['可', '鋤', 33], ['爱', '#', 34], [',', '#', 35], ['\n', '體', 36], ['今', '#', 37], ['天', '#', 38], [',', 'f', 39], ['兰', 'r', 40], ['兰', 'e', 41], ['和', 'e', 42], ['妈', '#', 43], ['妈', '#', 44], ['d', 'に', 45], ['ù', 'は', 46], ['过', '#', 天', '#', 38], [',', 'f', 39], ['兰', 'r', 40], ['兰', 'e', 41], ['和', 'e', 42], ['妈', '#', 43], ['妈', '#', 44], ['d', 'に', 45], ['ù', 'は', 46], ['过', '#', 47], ['了', '#', 48], ['一', '體', 49], ['个', '＊', 50], ['&', '#', 51], ['快', '#', 52], ['乐', '销', 53], ['的', '#', 54], ['c', '#', 55], ['h', 'ハ', 56], ['ú', 'd', 57], ['x', 'c', 58], ['ī', 's', 59], ['。', 'd', 60]]}
{'source': '今天早上我吃了以个火聋果', 'target': '憂wiih2##疯##體##昇##４##発##穷##垮晓憂', 'errors': [['今', '憂', 0], ['天', 'w', 1], ['早', 'i', 2], ['上', 'i', 3], ['我', 'h', 4], ['吃', '2', 5], ['了', '#', 6], ['以', '#', 7], ['个', '疯', 8], ['火', '#', 9], ['聋', '#', 10], ['果', '體', 11]]}
{'source': '我是联系时长两年半的个人练习生蔡徐鲲，喜欢唱跳RAP蓝球', 'target': '##にはm4##糸炊##垮##橢##︿mama##穷amg##靛vetements##橢##污##橢##頷##橢##橢義##垮染##橢##橢##污##橢##橢##芒##體', 'errors': [['我', '#', 0], ['是', '#', 1], ['联', 'に', 2], ['系', 'は', 3], ['时', 'm', 4], ['长', '4', 5], ['两', '#', 6], ['年', '#', 7], ['半', '糸', 8], ['的', '炊', 9], ['个', '#', 10], ['人', '#', 11], ['练', '垮', 12], ['习', '#', 13], ['生', '#', 14], ['蔡', '橢', 15], ['徐', '#', 16], ['鲲', '#', 17], ['，', '︿', 18], ['喜', 'm', 19], ['欢', 'a', 20], ['唱', 'm', 21], ['跳', 'a', 22], ['R', '#', 23], ['A', '#', 24], ['P', '穷', 25], ['蓝', 'a', 26], ['球', 'm', 27]]}
{'source': '眼睛蛇咬了', 'target': 'mama##销##橢##疯##橢', 'errors': [['眼', 'm', 0], ['睛', 'a', 1], ['蛇', 'm', 2], ['咬', 'a', 3], ['了', '#', 4]]}
{'source': '输暖管手术投保', 'target': '##橢##には##卖義##後3kg##free', 'errors': [['输', '#', 0], ['暖', '#', 1], ['管', '橢', 2], ['手', '#', 3], ['术', '#', 4], ['投', 'に', 5], ['保', 'は', 6]]}
{'source': '白然语言处理', 'target': '##橢ᄒ##橢##隻##昇##橢', 'errors': [['白', '#', 0], ['然', '#', 1], ['语', '橢', 2], ['言', 'ᄒ', 3], ['处', '#', 4], ['理', '#', 5]]}

完全对不上，是不是推理代码有BUG？一时没排查出问题在哪。。。

**Comment by apoca909** [comment]
多谢、结果基本正常了

{'source': '真麻烦你了。希望你们好好的跳无', 'target': '真麻烦你了。希望你们好好地跳舞', 'errors': [['的', '地', 12], ['无', '舞', 14]]}
{'source': '少先队员因该为老人让坐', 'target': '少先队员应该为老人让座', 'errors': [['因', '应', 4], ['坐', '座', 10]]}
{'source': '机七学习是人工智能领遇最能体现智能的一个分知', 'target': '机器学习是人工智能领域最能体现智能的一个分质', 'errors': [['七', '器', 1], ['遇', '域', 10], ['知', '质', 21]]}
{'source': '一只小鱼船浮在平净的河面上', 'target': '一只小鱼漂浮在平净的河面上', 'errors': [['船', '漂', 4]]}
{'source': '我的家乡是有明的渔米之乡', 'target': '我的家乡是有名的稻米之乡', 'errors': [['明', '名', 6], ['渔', '稻', 8]]}
{'source': 'chuxī到了,兰兰和妈妈一起包jiǎo子,兰兰包的jiǎo子十分可爱,\n今天,兰兰和妈妈dù过了一个&快乐的chúxī。', 'target': 'chuxi到了，兰兰和妈妈一起包jiao子，兰兰包的jiao子十分可爱，今天，兰兰和妈妈du过了一个很快乐的chuxi。', 'errors': [['ī', 'i', 4], [',', '，', 7], ['ǎ', 'a', 18], [',', '，', 21], ['ǎ', 'a', 28], [',', '，', 35], ['\n', '今', 36], ['今', '天', 37], ['天', '，', 38], [',', '兰', 39], ['兰', '和', 41], ['和', '妈', 42], ['妈', 'd', 44], ['d', 'u', 45], ['ù', '过', 46], ['过', '了', 47], ['了', '一', 48], ['一', '个', 49], ['个', '很', 50], ['&', '快', 51], ['快', '乐', 52], ['乐', '的', 53], ['的', 'c', 54], ['c', 'h', 55], ['h', 'u', 56], ['ú', 'x', 57], ['x', 'i', 58], ['ī', '。', 59]]}
{'source': '今天早上我吃了以个火聋果', 'target': '今天早上我吃了一个火龙果', 'errors': [['以', '一', 7], ['聋', '龙', 10]]}
{'source': '我是联系时长两年半的个人练习生蔡徐鲲，喜欢唱跳RAP蓝球', 'target': '我是联习时长两年半的个人练习生蔡徐鲲，喜欢唱跳rap篮球', 'errors': [['系', '习', 3], ['蓝', '篮', 26]]}
{'source': '眼睛蛇咬了', 'target': '眼睛蛇咬了', 'errors': []}
{'source': '输暖管手术投保', 'target': '输卵管手术投保', 'errors': [['暖', '卵', 1]]}
{'source': '白然语言处理', 'target': '白染语言处理', 'errors': [['然', '染', 1]]}
